snow/engine/snowman/event/queue_test.go

```diff
@@ -0,0 +1,294 @@
+			return nil
+		},
+		cancel: func(context.Context) error {
+			return errUnexpectedInvocation
+		},
+	}
+
+	tests := []struct {
+		name          string
+		queue         *Queue[int]
+		dependencies  []int
+		shouldExecute bool
+		expectedLen   int
+		expectedQueue *Queue[int]
+	}{
+		{
+			name:          "no dependencies",
+			queue:         NewQueue[int](),
+			dependencies:  nil,
+			shouldExecute: true,
+			expectedLen:   0,
+			expectedQueue: NewQueue[int](),
+		},
+		{
+			name:          "one dependency",
+			queue:         NewQueue[int](),
+			dependencies:  []int{1},
+			shouldExecute: false,
+			expectedLen:   1,
+			expectedQueue: &Queue[int]{
+				jobs: map[int][]*job[int]{
+					1: {
+						{
+							dependencies: set.Of(1),
+							job:          userJob,
+						},
+					},
+				},
+			},
+		},
+		{
+			name:          "two dependencies",
+			queue:         NewQueue[int](),
+			dependencies:  []int{1, 2},
+			shouldExecute: false,
+			expectedLen:   2,
+			expectedQueue: &Queue[int]{
+				jobs: map[int][]*job[int]{
+					1: {
+						{
+							dependencies: set.Of(1, 2),
+							job:          userJob,
+						},
+					},
+					2: {
+						{
+							dependencies: set.Of(1, 2),
+							job:          userJob,
+						},
+					},
+				},
+			},
+		},
+		{
+			name: "additional dependency",
+			queue: &Queue[int]{
+				jobs: map[int][]*job[int]{
+					1: {
+						{
+							dependencies: set.Of(1),
+							job:          userJob,
+						},
+					},
+				},
+			},
+			dependencies:  []int{1},
+			shouldExecute: false,
+			expectedLen:   1,
+			expectedQueue: &Queue[int]{
+				jobs: map[int][]*job[int]{
+					1: {
+						{
+							dependencies: set.Of(1),
+							job:          userJob,
+						},
+						{
+							dependencies: set.Of(1),
+							job:          userJob,
+						},
+					},
+				},
+			},
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			require := require.New(t)
+
+			// Reset the variable between tests
+			calledExecute = false
+
+			require.NoError(test.queue.Register(context.Background(), userJob, test.dependencies...))
+			require.Equal(test.expectedLen, test.queue.NumDependencies())
+			require.Equal(test.shouldExecute, calledExecute)
+			require.Equal(test.expectedQueue, test.queue)
+		})
+	}
+}
+
+func TestQueue_Fulfill(t *testing.T) {
+	var calledExecute bool
+	userJob := &testJob{
+		execute: func(context.Context) error {
+			if calledExecute {
+				return errDuplicateInvocation
+			}
+			calledExecute = true
+			return nil
+		},
+		cancel: func(context.Context) error {
+			return errUnexpectedInvocation
+		},
+	}
+
+	tests := []struct {
+		name          string
+		queue         *Queue[int]
+		shouldExecute bool
+		expectedQueue *Queue[int]
+	}{
+		{
+			name:          "no jobs",
+			queue:         NewQueue[int](),
+			shouldExecute: false,
+			expectedQueue: NewQueue[int](),
+		},
+		{
+			name:          "single dependency",
+			queue:         newQueueWithJob(t, userJob, false, 1),
```

Comment:
The use of `1` vs `2` is important but not obvious until later. I think you should introduce constants before the test table as it currently requires reading the test loop to be understood:

```go
const (
  depToFulfill = iota
  depToNeglect
)

...

test.queue.Fulfill(..., depToFulfil)
```

Same applies to the `Abandon()` test.